### PR TITLE
Books routes

### PR DESCRIPTION
**Pull request by mnmallea**
## Summary
- Implemented index route for Books.
- Implemented show route for Books.
- Both routes requires an authenticated user.
- Index route allows to search books by author, genre or title.

## Trello card
[Link](https://trello.com/c/cquuRpQI/21-crear-rutas-index-y-show-para-book)

### CODE REVIEW

File: app/serializers/book_preview_serializer.rb [hunk 1]
```diff
@@ -0,0 +1,6 @@
+class BookPreviewSerializer < ActiveModel::Serializer
+  attributes :id, :author, :title, :publisher, :year, :genre, :created_at, :updated_at
+  attribute(:image) do
```
Comment (leoncalermo) [inline app/serializers/book_preview_serializer.rb]: instead this i would use this
```
def image
  { url: object.image } 
end
```
and in attributes above i would add the `:image` at the end of the line

File: app/serializers/actual_rent_serializer.rb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+class ActualRentSerializer < ActiveModel::Serializer
+  attributes :id, :book_id, :user_id
+  attribute :starting_date, key: :from
```
Comment (leoncalermo) [inline app/serializers/actual_rent_serializer.rb]: instead this i would use this
```
def from
  object.starting_date
end
```
and add `:from` in the above attributes line

File: app/serializers/actual_rent_serializer.rb [hunk 3]
```diff
@@ -0,0 +1,5 @@
...
+  attribute :ending_date, key: :to
```
Comment (leoncalermo) [inline app/serializers/actual_rent_serializer.rb]: same

File: app/serializers/book_serializer.rb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+class BookSerializer < ActiveModel::Serializer
+  attributes :id, :author, :title, :publisher, :year, :genre
+  attribute :image, key: :image_url
```
Comment (leoncalermo) [inline app/serializers/book_serializer.rb]: same

File: app/controllers/books_controller.rb [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+  def show
+    if book
+      render json: book
+    else
+      render json: {}, status: :not_found
+    end
+  end
+
+  def book
+    @book ||= Book.find params[:id]
```
Comment (leoncalermo) [inline app/controllers/books_controller.rb]: we try to use the  parenthesis in the functions  for easy reading

File: app/controllers/books_controller.rb [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+  include Wor::Paginate
+
+  before_action :authenticate_user!
+
+  def index
+    render_paginated Book::Reducer.apply(params), each_serializer: BookPreviewSerializer
+  end
+
+  def show
+    if book
```
Comment (leoncalermo) [inline app/controllers/books_controller.rb]: we try to not use this way and even we try to not use else too hahah
in this case i would do something like this 
```
def show
  return render json: book if book.present?
  head :not_found
end
```

File: app/models/book.rb [hunk 7]
```diff
@@ -14,4 +14,9 @@ class Book < ApplicationRecord
   )
 
   has_many :rents, dependent: :destroy
+
+  def actual_rent
+    rents.find_by(':current_date between starting_date and ending_date',
```
Comment (leoncalermo) [inline app/models/book.rb:19]: this works?!? it's amazing 😍

**Review by leoncalermo — CHANGES_REQUESTED** [review]

File: app/models/book.rb (hunk 7 above)
Comment (mnmallea) [reply app/models/book.rb:19]: yes :smile:

**Review by leoncalermo — APPROVED** [review]

**Pull request merged**